### PR TITLE
Add visible in listings flag to product

### PR DESCRIPTION
**Pull request by IKarbowiak**
- Add `visible_in_listings` to `Product` model
- Update Product type with `visible_in_listings` field
- Update `productCreate` and `productUpdate` inptus with `visibleInListings`
- Update csv export with`visible_in_listings` field
- Update populatedb - 10% of generated products will have `visible_in_listings` set to False
- Update top-level `products` and `variants` resolver to exclude products with `visible_in_listings=false` (for users without permission)
- Update `Category.products` to exclude products with `visible_in_listings=false` (for users without permission)
- Hide attributes that belong to hidden products

⚠️ 
We shouldn't merge it before the frontend part is not ready.

# Impact

* [x] New migrations
* [x] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [x] The changes are tested
* [x] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/graphql/product/mutations/products.py [hunk 1]
```diff
@@ -540,6 +540,9 @@ class ProductInput(graphene.InputObjectType):
             "Note: this field is only used if a product doesn't use variants."
         )
     )
+    visible_in_listings = graphene.Boolean(
+        description="Determines if product is visible in product listings."
```
Comment (korycins) [inline saleor/graphql/product/mutations/products.py:544]: Don't we want to have default value here?

File: saleor/graphql/product/tests/test_product.py [hunk 2]
```diff
@@ -301,6 +310,64 @@ def test_fetch_unavailable_products(user_api_client, product):
...
+
+    product_count = Product.objects.count()
+
+    # when
+    response = user_api_client.post_graphql(QUERY_FETCH_ALL_PRODUCTS)
+
+    # then
+    content = get_graphql_content(response)
+    product_data = content["data"]["products"]["edges"]
+    assert len(product_data) == product_count - 1
```
Comment (korycins) [inline saleor/graphql/product/tests/test_product.py:328]: Don't you think that we should compare here at least the returned IDs? Or confirm that we didn't get product with `visible_in_listings == False`
Something like :
`assert id_of_product_not_visible not in product_data[IDs]` ?

**Review by korycins — APPROVED** [review]

File: saleor/graphql/product/mutations/products.py (hunk 1 above)
Comment (IKarbowiak) [reply saleor/graphql/product/mutations/products.py:544]: The default value is in the model field definition. I think it's enough. :)

File: saleor/graphql/product/tests/test_attributes.py [hunk 3]
```diff
@@ -1982,6 +1984,166 @@ def test_filter_attributes_by_global_id_list(api_client, attribute_list):
...
+    )["data"]["attributes"]["edges"]
+
+    # then
+    assert len(attributes) == attribute_count - 1
+    assert weight_attribute.slug not in {
+        attribute["node"]["slug"] for attribute in attributes
+    }
+
+
+def test_filter_attributes_in_category_by_staff_with_perm(
```
Comment (fowczarek) [inline saleor/graphql/product/tests/test_attributes.py:2027]: Lack of test from `app_api_client`.

File: saleor/graphql/product/tests/test_category.py [hunk 4]
```diff
@@ -95,6 +102,77 @@ def test_category_query_error_when_no_param(
...
+
+    # when
+    response = staff_api_client.post_graphql(QUERY_CATEGORY, variables=variables)
+
+    # then
+    content = get_graphql_content(response, ignore_errors=True)
+    assert len(content["data"]["category"]["products"]["edges"]) == product_count - 1
+
+
+def test_query_category_product_only_visible_in_listings_as_staff_with_perm(
```
Comment (fowczarek) [inline saleor/graphql/product/tests/test_category.py:151]: Lack of test from app_api_client.

File: saleor/graphql/product/types/products.py [hunk 5]
```diff
@@ -499,6 +499,7 @@ class Meta:
             "seo_title",
             "updated_at",
             "weight",
+            "visible_in_listings",
```
Comment (fowczarek) [inline saleor/graphql/product/types/products.py:502]: This field should be available to customers?

File: saleor/graphql/product/tests/test_product.py [hunk 6]
```diff
@@ -301,6 +311,68 @@ def test_fetch_unavailable_products(user_api_client, product):
...
+    products_ids = [product["node"]["id"] for product in product_data]
+    assert graphene.Node.to_global_id("Product", product_list[0].pk) not in products_ids
+
+
+def test_fetch_all_products_visible_in_listings_by_staff_with_perm(
```
Comment (fowczarek) [inline saleor/graphql/product/tests/test_product.py:334]: Lack of test from app_api_client.

File: saleor/graphql/product/tests/test_variant.py [hunk 7]
```diff
@@ -966,6 +966,75 @@ def test_fetch_all_variants_anonymous_user(
...
+
+    product_count = Product.objects.count()
+
+    # when
+    data = _fetch_all_variants(staff_api_client)
+
+    assert data["totalCount"] == product_count - 1
+
+
+def test_product_variants_visible_in_listings_by_staff_with_perm(
```
Comment (fowczarek) [inline saleor/graphql/product/tests/test_variant.py:1021]: Lack of test from app_api_client.

**Review by fowczarek — APPROVED** [review]

File: saleor/graphql/product/types/products.py (hunk 5 above)
Comment (IKarbowiak) [reply saleor/graphql/product/types/products.py:502]: I did it in the same way as `is_published` field. But maybe both of these fields should be protected be some permission.  What do you think @patrys ?

Comment (fowczarek) [reply saleor/graphql/product/types/products.py:502]: We should leave this field as public. In some cases you want to hide some navigation depends on this field.

**Comment by IKarbowiak** [comment]
Changes in #6060

**Pull request closed without merging**